skafka/src/main/scala/com/evolutiongaming/skafka/consumer/RebalanceCallback.scala

```diff
@@ -0,0 +1,267 @@
+    )
+
+  // TODO using Offset.unsafe to avoid requiring ApplicativeThrowable as in Offset.of
+  //  also it looks like consumer.position cannot return invalid offset
+  def position[F[_]](tp: TopicPartition): RebalanceCallback[F, Offset] =
+    ConsumerStep(c => Offset.unsafe(c.position(tp.asJava)))
+
+  def position[F[_]](tp: TopicPartition, timeout: FiniteDuration): RebalanceCallback[F, Offset] =
+    ConsumerStep(c => Offset.unsafe(c.position(tp.asJava, timeout.asJava)))
+
+  def seek[F[_]](partition: TopicPartition, offset: Offset): RebalanceCallback[F, Unit] =
+    ConsumerStep(c => c.seek(partition.asJava, offset.value))
+
+  def seek[F[_]](partition: TopicPartition, offsetAndMetadata: OffsetAndMetadata): RebalanceCallback[F, Unit] =
+    ConsumerStep(c => c.seek(partition.asJava, offsetAndMetadata.asJava))
+
+  def seekToBeginning[F[_]](partitions: Nes[TopicPartition]): RebalanceCallback[F, Unit] =
+    ConsumerStep(c => c.seekToBeginning(partitions.asJava))
+
+  def seekToEnd[F[_]](partitions: Nes[TopicPartition]): RebalanceCallback[F, Unit] =
+    ConsumerStep(c => c.seekToEnd(partitions.asJava))
+
+  def subscription[F[_]]: RebalanceCallback[F, Set[Topic]] = ConsumerStep(_.subscription().asScala.toSet)
+
+  private[consumer] def run[F[_]: Monad: ToTry, A](
+    rc: RebalanceCallback[F, A],
+    consumer: ConsumerJ[_, _]
+  ): Try[A] = {
+
+    // TODO @tailrec failed to convert map/flatMap to patmat in 5 minutes, was getting scalac errors about type mismatches
+    def loop[A1](rc: RebalanceCallback[F, A1]): Try[A1] = {
+      rc match {
+        case Pure(a)                => Try(a)
+        case MapStep(f, parent)     => loop(parent).map(f)
+        case FlatMapStep(f, parent) => loop(parent).flatMap { a => loop(f(a)) }
+        case LiftStep(fa)           => fa.toTry
+        case ConsumerStep(f)        => Try { f(consumer) }
+      }
+    }
+
+    loop(rc)
+  }
+
+  private final case class Pure[F[_], A](a: A) extends RebalanceCallback[F, A] {
+    def mapK[G[_]](fg: F ~> G): RebalanceCallback[G, A] = Pure[G, A](a)
+  }
+
+  private final case class MapStep[F[_], A, B](f: A => B, parent: RebalanceCallback[F, A])
+      extends RebalanceCallback[F, B] {
+    def mapK[G[_]](fg: F ~> G): RebalanceCallback[G, B] = MapStep[G, A, B](f, parent.mapK(fg))
+  }
+
+  private final case class FlatMapStep[F[_], A, B](f: A => RebalanceCallback[F, B], parent: RebalanceCallback[F, A])
+      extends RebalanceCallback[F, B] {
+    def mapK[G[_]](fg: F ~> G): RebalanceCallback[G, B] = FlatMapStep[G, A, B](f andThen (_.mapK(fg)), parent.mapK(fg))
+  }
+
+  private final case class LiftStep[F[_], A](fa: F[A]) extends RebalanceCallback[F, A] {
+    def mapK[G[_]](fg: F ~> G): RebalanceCallback[G, A] = LiftStep[G, A](fg(fa))
+  }
+
+  private final case class ConsumerStep[F[_], A](f: ConsumerJ[_, _] => A) extends RebalanceCallback[F, A] {
+    def mapK[G[_]](fg: F ~> G): RebalanceCallback[G, A] = ConsumerStep[G, A](f)
```

Comment:
does this compile?

Reply:
yes and it works

Reply:
as agreed - extracted to extension method in bbd8c114c0e5a9efeb82e1f338b7055c26eb57b7